src/ngx_http_lua_shdict.c

```diff
@@ -2846,4 +2850,72 @@ ngx_http_lua_ffi_shdict_flush_all(ngx_shm_zone_t *zone)
 #endif /* NGX_LUA_NO_FFI_API */
 
 
+static int
+ngx_http_lua_shdict_get_stats(lua_State *L)
+{
+    ngx_http_lua_shdict_ctx_t   *ctx;
+    ngx_shm_zone_t              *zone;
+    ngx_slab_stat_t             *stats;
+    int                          n;
+    int                          n_slots;
+    int                          slot;
+    int                          size;
+
+    n = lua_gettop(L);
+
+    if (n != 1) {
+        return luaL_error(L, "expecting 1 argument, "
+                          "but saw %d", n);
+    }
+
+    luaL_checktype(L, 1, LUA_TTABLE);
+
+    zone = ngx_http_lua_shdict_get_zone(L, 1);
+    if (zone == NULL) {
+        return luaL_error(L, "bad user data for the ngx_shm_zone_t pointer");
+    }
+
+    ctx = zone->data;
+
+    ngx_shmtx_lock(&ctx->shpool->mutex);
+
+    n_slots = ngx_pagesize_shift - ctx->shpool->min_shift;
+    lua_createtable(L, n_slots, 0);
```

Comment:
We cannot really hold the lock while calling these `lua_XXX` C API functions since the latter may introduce GC steps that may call `__gc` meta methods which in turn call into shdict methods, thus leading to deadlocks (we're about to retire all the existing CFunction-based Lua API functions in ngx_lua module to address similar issues).

Reply:
Also, it's better to allow the caller to supply this Lua table so that we do not have to create a brand new Lua table upon every invocation of this API function, saving dynamic allocation and GC overhead.

Reply:
We should implement this as an pure C function instead and make a FFI Lua wrapper around it.